text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,499 @@
+set of functions which read, write, and propagate data.
+
+**SetBaggage(context, key, value) -> context**  
+To record the distributed state of an aspect, the Baggage API provides a 
+function which takes a context, a key, and a value as input, and returns an 
+updated context which contains the new value.
+
+**GetBaggage(context, key) -> value**  
+To access the distributed state of an aspect, the Baggage API provides a 
+function which takes a context and a key as input, and returns a value.
+
+**RemoveBaggage(context, key) -> context**  
+To delete distributed state from an aspect, the Baggage API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**ClearBaggage(context) -> context**  
+To avoid sending baggage to an untrusted downstream process, the Baggage API 
+provides a function remove all baggage from a context.
+
+**GetBaggageHTTPPropagator() -> (HTTPExtract, HTTPInject)**  
+To deserialize the state of the system sent from the the prior upstream process, 
+and to serialize the the current state of the system and send it to the next 
+downstream process, the Baggage API provides a function which returns a 
+baggage-specific implementation of the HTTPExtract and HTTPInject functions.
+
+# Context Propagation
+
+## Context API
+
+Aspects access data in-process using a shared context object. Each aspect uses 
+its own namespaced set of keys in the context, containing all of the data for 
+that cross-cutting concern.
+
+**SetValue(context, key, value) -> context**  
+To record the local state of an aspect, the Context API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**GetValue(context, key) -> value**  
+To access the local state of an aspect, the Context API provides a function 
+which takes a context and a key as input, and returns a value.
+
+### Optional: Automated Context Management
+When possible, the OpenTelemetry context should automatically be associated 
+with the program execution context. Note that some languages do not provide any 
+facility for setting and getting a current context. In these cases, the user is 
+responsible for managing the current context.  
+
+**SetCurrent(context)**  
+To associate a context with program execution, the Context API provides a 
+function which takes a Context.
+
+**GetCurrent() -> context**  
+To access the context associated with program execution, the Context API 
+provides a function which takes no arguments and returns a Context.
+
+
+## Propagation API
+
+Aspects send their state to downstream processes via propagators: 
+functions which read and write context into RPC requests. Each aspect creates a 
+set of propagators for every type of supported medium - currently only HTTP 
+requests.
+
+**HTTPExtract(context, request) -> context**  
```

Comment:
in practice this can get a bit nasty, specifically in the representation of a request.

For example, in Python there are multiple different web frameworks, each with a slightly different implementation of a request. As such, it would be required for each integration to implement some adapter class, which can become quite tedious as HTTP provides multiple different objects of storage.

That said, I assume the value here is to enable propagation via a broader set of methods, including things not possible today like body and query parameters.

My two cents is maybe argue for that big of a change for extractors in a separate OTEP. I think the only thing that we need today to implement baggage and all other protocols is the ability to receive the whole set of headers. I know @tsloughter had a good idea there to just pass in a header iterator for extract, and retrieve one for inject. My one amendment to that idea is to pass in a header object with some way to iterate keys.

Reply:
Sorry, I should update this to say "headers" instead of "request," it was not my intention to actually change this detail.